File: apps/api/app/schemas/talent_market_map.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


SegmentType = Literal["primary", "adjacent", "exploratory"]
PriorityLevel = Literal["high", "medium", "low"]
ClosenessLevel = Literal["high", "medium", "low"]
ImpactLevel = Literal["high", "medium", "low"]
ConfidenceLevel = Literal["high", "medium", "low"]
MapStatus = Literal["draft", "generated", "updated", "archived"]
MarketAssessment = Literal["broad", "moderate", "narrow", "very_narrow"]
SegmentCoverage = Literal[
    "not_started", "in_progress", "partially_covered", "covered", "discarded"
]
CompanyCoverage = Literal[
    "not_reviewed",
    "in_review",
    "partially_covered",
    "covered",
    "no_relevant_candidates",
    "discarded",
]
RecommendationStatus = Literal["suggested", "accepted", "rejected"]


# --- Sub-entidades: read -----------------------------------------------------


class MarketSegmentRead(BaseModel):
    id: int
    market_map_id: int
    name: str
    segment_type: SegmentType
    description: str | None
    priority: PriorityLevel
    coverage_status: SegmentCoverage
    rationale: str | None
    sort_order: int
    ai_suggested: bool
    created_at: datetime
    updated_at: datetime
    candidate_count: int = 0  # derivado

    model_config = {"from_attributes": True}


class TargetCompanyRead(BaseModel):
    id: int
    market_map_id: int
    segment_id: int | None
    name: str
    industry: str | None
    priority: PriorityLevel
    rationale: str | None
    coverage_status: CompanyCoverage
    notes: str | None
    ai_suggested: bool
    created_at: datetime
    updated_at: datetime
    candidates_identified: int = 0  # derivado
    candidates_evaluated: int = 0  # derivado
    high_fit_candidates: int = 0  # derivado

    model_config = {"from_attributes": True}


class EquivalentRoleRead(BaseModel):
    id: int
    market_map_id: int
    title: str
    seniority: str | None
    closeness: ClosenessLevel
    rationale: str | None
    priority: PriorityLevel
    industries: list[str] = Field(default_factory=list)
    ai_suggested: bool
    created_at: datetime
    updated_at: datetime
    candidate_count: int = 0  # derivado

    model_config = {"from_attributes": True}


class MarketGapRead(BaseModel):
    id: int
    market_map_id: int
    title: str
    frequency: int
    total_evaluated: int
    impact: ImpactLevel
    evidence: str | None
    recommendation: str | None
    detected_at: datetime

    model_config = {"from_attributes": True}


class RecalibrationRecommendationRead(BaseModel):
    id: int
    market_map_id: int
    title: str
    reason: str
    expected_impact: str | None
    confidence: ConfidenceLevel
    status: RecommendationStatus
    generated_by: Literal["rules", "ai"]
    acted_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Cobertura derivada ------------------------------------------------------


class CoverageStats(BaseModel):
    candidates_identified: int
    candidates_loaded: int
    candidates_evaluated: int
    high_fit: int
    medium_fit: int
    low_fit: int
    discarded: int
    shortlisted: int
    target_companies_total: int
    target_companies_reviewed: int
    target_companies_pending: int
    industries_covered: int
    coverage_pct: int  # 0..100


# --- Map read ---------------------------------------------------------------


class TalentMarketMapRead(BaseModel):
    id: int
    search_mandate_id: int
    position_spec_id: int | None
    status: MapStatus
    executive_summary: str | None
    executive_summary_for_client: str | None
    market_assessment: MarketAssessment | None
    generated_by_model: str | None
    prompt_version: str | None
    generated_at: datetime | None
    created_at: datetime
    updated_at: datetime
    segments: list[MarketSegmentRead] = Field(default_factory=list)
    companies: list[TargetCompanyRead] = Field(default_factory=list)
    equivalent_roles: list[EquivalentRoleRead] = Field(default_factory=list)
    gaps: list[MarketGapRead] = Field(default_factory=list)
    recommendations: list[RecalibrationRecommendationRead] = Field(default_factory=list)
    coverage: CoverageStats


# --- Payloads CRUD ---------------------------------------------------------


class MapUpdatePayload(BaseModel):
    executive_summary: str | None = None
    executive_summary_for_client: str | None = None
    market_assessment: MarketAssessment | None = None
    status: MapStatus | None = None


class SegmentCreatePayload(BaseModel):
    name: str = Field(..., max_length=200)
    segment_type: SegmentType
    description: str | None = None
    priority: PriorityLevel = "medium"
    coverage_status: SegmentCoverage = "not_started"
    rationale: str | None = None


class SegmentUpdatePayload(BaseModel):
    name: str | None = Field(default=None, max_length=200)
    segment_type: SegmentType | None = None
    description: str | None = None
    priority: PriorityLevel | None = None
    coverage_status: SegmentCoverage | None = None
    rationale: str | None = None
    sort_order: int | None = None


class SegmentReorderPayload(BaseModel):
    ordered_ids: list[int] = Field(..., min_length=1)


class CompanyCreatePayload(BaseModel):
    name: str = Field(..., max_length=200)
    industry: str | None = Field(default=None, max_length=120)
    segment_id: int | None = None
    priority: PriorityLevel = "medium"
    rationale: str | None = None
    coverage_status: CompanyCoverage = "not_reviewed"
    notes: str | None = None


class CompanyUpdatePayload(BaseModel):
    name: str | None = Field(default=None, max_length=200)
    industry: str | None = Field(default=None, max_length=120)
    segment_id: int | None = None
    priority: PriorityLevel | None = None
    rationale: str | None = None
    coverage_status: CompanyCoverage | None = None
    notes: str | None = None


class EquivalentRoleCreatePayload(BaseModel):
    title: str = Field(..., max_length=200)
    seniority: str | None = Field(default=None, max_length=80)
    closeness: ClosenessLevel = "medium"
    rationale: str | None = None
    priority: PriorityLevel = "medium"
    industries: list[str] = Field(default_factory=list)


class EquivalentRoleUpdatePayload(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    seniority: str | None = Field(default=None, max_length=80)
    closeness: ClosenessLevel | None = None
    rationale: str | None = None
    priority: PriorityLevel | None = None
    industries: list[str] | None = None


class MapCandidateRead(BaseModel):
    candidate_id: int
    full_name: str
    current_company: str | None
    current_position: str | None
    evaluation_score: int | None
    score_category: str | None
    auto_company_id: int | None
    segment_id: int | None
    target_company_id: int | None
    equivalent_role_id: int | None


class CandidateAssignPayload(BaseModel):
    segment_id: int | None = None
    target_company_id: int | None = None
    equivalent_role_id: int | None = None


class RecommendationDecisionPayload(BaseModel):
    status: Literal["accepted", "rejected"]


class GenerateMapPayload(BaseModel):
    """Opcionalmente pisar entidades AI suggeridas; default no pisa edits manuales."""

    overwrite_ai_suggested: bool = True
    overwrite_manual: bool = False

File: apps/web/types/talent-market-map.ts
// Tipos del Talent Market Map. Espejo de app/schemas/talent_market_map.py.

export type SegmentType = "primary" | "adjacent" | "exploratory";
export type PriorityLevel = "high" | "medium" | "low";
export type ClosenessLevel = "high" | "medium" | "low";
export type ImpactLevel = "high" | "medium" | "low";
export type ConfidenceLevel = "high" | "medium" | "low";
export type MapStatus = "draft" | "generated" | "updated" | "archived";
export type MarketAssessment = "broad" | "moderate" | "narrow" | "very_narrow";

export type SegmentCoverage =
  | "not_started"
  | "in_progress"
  | "partially_covered"
  | "covered"
  | "discarded";

export type CompanyCoverage =
  | "not_reviewed"
  | "in_review"
  | "partially_covered"
  | "covered"
  | "no_relevant_candidates"
  | "discarded";

export type RecommendationStatus = "suggested" | "accepted" | "rejected";

export type MarketSegment = {
  id: number;
  market_map_id: number;
  name: string;
  segment_type: SegmentType;
  description: string | null;
  priority: PriorityLevel;
  coverage_status: SegmentCoverage;
  rationale: string | null;
  sort_order: number;
  ai_suggested: boolean;
  created_at: string;
  updated_at: string;
  candidate_count: number;
};

export type TargetCompany = {
  id: number;
  market_map_id: number;
  segment_id: number | null;
  name: string;
  industry: string | null;
  priority: PriorityLevel;
  rationale: string | null;
  coverage_status: CompanyCoverage;
  notes: string | null;
  ai_suggested: boolean;
  created_at: string;
  updated_at: string;
  candidates_identified: number;
  candidates_evaluated: number;
  high_fit_candidates: number;
};

export type EquivalentRole = {
  id: number;
  market_map_id: number;
  title: string;
  seniority: string | null;
  closeness: ClosenessLevel;
  rationale: string | null;
  priority: PriorityLevel;
  industries: string[];
  ai_suggested: boolean;
  created_at: string;
  updated_at: string;
  candidate_count: number;
};

export type MarketGap = {
  id: number;
  market_map_id: number;
  title: string;
  frequency: number;
  total_evaluated: number;
  impact: ImpactLevel;
  evidence: string | null;
  recommendation: string | null;
  detected_at: string;
};

export type RecalibrationRecommendation = {
  id: number;
  market_map_id: number;
  title: string;
  reason: string;
  expected_impact: string | null;
  confidence: ConfidenceLevel;
  status: RecommendationStatus;
  generated_by: "rules" | "ai";
  acted_at: string | null;
  created_at: string;
};

export type CoverageStats = {
  candidates_identified: number;
  candidates_loaded: number;
  candidates_evaluated: number;
  high_fit: number;
  medium_fit: number;
  low_fit: number;
  discarded: number;
  shortlisted: number;
  target_companies_total: number;
  target_companies_reviewed: number;
  target_companies_pending: number;
  industries_covered: number;
  coverage_pct: number;
};

export type TalentMarketMap = {
  id: number;
  search_mandate_id: number;
  position_spec_id: number | null;
  status: MapStatus;
  executive_summary: string | null;
  executive_summary_for_client: string | null;
  market_assessment: MarketAssessment | null;
  generated_by_model: string | null;
  prompt_version: string | null;
  generated_at: string | null;
  created_at: string;
  updated_at: string;
  segments: MarketSegment[];
  companies: TargetCompany[];
  equivalent_roles: EquivalentRole[];
  gaps: MarketGap[];
  recommendations: RecalibrationRecommendation[];
  coverage: CoverageStats;
};

// --- Etiquetas en español -------------------------------------------------

export const MAP_STATUS_LABELS: Record<MapStatus, string> = {
  draft: "Borrador",
  generated: "Generado",
  updated: "Actualizado",
  archived: "Archivado",
};

export const MARKET_ASSESSMENT_LABELS: Record<MarketAssessment, string> = {
  broad: "Mercado amplio",
  moderate: "Mercado moderado",
  narrow: "Mercado estrecho",
  very_narrow: "Mercado muy estrecho",
};

export const SEGMENT_TYPE_LABELS: Record<SegmentType, string> = {
  primary: "Principal",
  adjacent: "Adyacente",
  exploratory: "Exploratorio",
};

export const PRIORITY_LABELS: Record<PriorityLevel, string> = {
  high: "Alta",
  medium: "Media",
  low: "Baja",
};

export const CLOSENESS_LABELS: Record<ClosenessLevel, string> = {
  high: "Cercanía alta",
  medium: "Cercanía media",
  low: "Cercanía baja",
};

export const IMPACT_LABELS: Record<ImpactLevel, string> = {
  high: "Impacto alto",
  medium: "Impacto medio",
  low: "Impacto bajo",
};

export const SEGMENT_COVERAGE_LABELS: Record<SegmentCoverage, string> = {
  not_started: "Sin iniciar",
  in_progress: "En progreso",
  partially_covered: "Parcialmente cubierto",
  covered: "Cubierto",
  discarded: "Descartado",
};

export const COMPANY_COVERAGE_LABELS: Record<CompanyCoverage, string> = {
  not_reviewed: "Sin revisar",
  in_review: "En revisión",
  partially_covered: "Parcialmente cubierta",
  covered: "Cubierta",
  no_relevant_candidates: "Sin candidatos relevantes",
  discarded: "Descartada",
};

export const RECOMMENDATION_STATUS_LABELS: Record<RecommendationStatus, string> = {
  suggested: "Sugerida",
  accepted: "Aceptada",
  rejected: "Descartada",
};
